Rolling PR for bookmark merging and syncing

linacambridge opened the pull request:
This PR adds a `BookmarksStore` for syncing, separate tables for storing incoming bookmarks and staging outgoing ones, and integrates Dogear for merging.

Changes to Desktop's schema:

* We just stash incoming URLs directly into `moz_places`. since we'll be merging them, anyway, and expiration should take care of any stale ones. This avoids an extra table and index over the same data.
* Ditto for `moz_tags`.
* No separate `meta` table. We'll already use `moz_meta` to store the ~sync ID and~ last sync time.
* `mirror.items` -> `moz_bookmarks_synced`, `mirror.structure` -> `moz_bookmarks_synced_structure`, `mirror.tags` -> `moz_bookmarks_synced_tag_relation`.
* `mergeStates` -> `mergedTree`.
* I tried to scrub references to "mirror" and iOS in the comments where I could, since we've diverged pretty far from iOS' original implementation, and `moz_bookmarks_synced` isn't a true mirror like the logins mirror is.
* No observer tables or bookkeeping, since we don't fire those.

Closes #575. Closes #576. Closes #728. Depends on #786.

mhammond commented:
>     * Why do we allow initializing the schema twice (`test_create_schema_twice`)? For migration? Temp tables and triggers don't use `IF NOT EXISTS`, so I was wondering if this is deliberate.

Some migrations just re-run that sql so new tables/indexes are correctly added. IIUC, there's no requirement that it can run twice on the same connection, but that was just the simplest way of writing the test. We can probably adjust that test easily enough.

linacambridge commented:
> IIUC, there's no requirement that it can run twice on the same connection, but that was just the simplest way of writing the test.

That makes sense, thanks! I think it's OK to leave as is, I was just curious if there was a reason I was missing besides migrations. 😊

linacambridge commented on components/places/sql/create_sync_temp_tables.sql:
```diff
@@ -0,0 +1,134 @@
+-- Stores local IDs for items to reupload. Removing an
+-- ID from this table bumps its local change counter, so, unlike weak uploads,
+-- we *will* reupload the item on the next sync if the current sync fails.
+-- This is used to ensure that all bookmarks with the same URL have the same keyword (bug 1328737).
```
I think we can remove this complexity for keywords. Since we're not exposing any UI to edit them (IOW, keyword changes will only be remote, not local), we can just round-trip them, and store them with the bookmark instead of maintaining keyword-URL associations.

linacambridge commented:
Thanks, @mhammond, this is a great start! And thanks for adding the test harness, building on that was a breeze. I made a few changes and wired up Dogear, so we have merging working now!

* `BookmarkNode` and `BookmarkTree` are part of Dogear, so we don't have to implement it ourselves! 🎉  I'm really sorry I provided misleading and insufficient guidance. In [this comment](https://github.com/mozilla/dogear/issues/24#issuecomment-469102343), I meant we should copy how my Rust Desktop patch does it, not how Desktop's `SyncedBookmarksMirror.jsm` does it today.
* We also don't use `PseudoTree` (or `inflate_tree`) anymore.
* `Merger` is what I intended for `BookmarksStore` to be, so I moved your methods into it. (I think this also answers your question about "or is dogear the merger, and this something else?" Dogear is the merger; `BookmarksStore` is the glue between Sync and Dogear). Dogear's `Store` trait defines `fetch_local_tree`, `fetch_remote_tree`, and `apply`; we supply the implementation. I just copy-pasted your SQL. 😄

Answering some of your questions:

* "should we have an "unknown" for SyncedBookmarkKind?": I'm leaning no, because we can't meaningfully sync or merge items we don't support.
* "XXX - need tags!": The quick-and-dirty fix for tags I mentioned earlier is associate them with bookmarks, not URLs. However, since we do plan to support tags, I think it's worth doing it right, and copying [this patch](https://bugzilla.mozilla.org/show_bug.cgi?id=1039200). Desktop's "tags as bookmarks" is a complication that we should definitely leave behind. Keywords can be associated with bookmarks, though, since we only plan to round-trip them.
* "XXX - we appear to have change counter issues with the tree?": We aren't bumping the change counter of the parent when we insert, fixed in #742.

linacambridge commented on components/places/src/bookmark_sync/record.rs:
```diff
@@ -0,0 +1,244 @@
+            }
+            _ => {}
+        }
+        // TODO: We don't know how to round-trip other kinds. For now, just
```
Is it possible to return a custom error from `deserialize`? I couldn't figure out how to unify our errors with `D::Error`.

mhammond replied:
I think we just need `Err(de::Error::custom(format!("unknown bookmark record type '{}'", raw.kind)))` or similar. sync15's error.rs maps serde errors to ours, so I *think* that will end up as a `JsonError`, which isn't quite correct, but probably good enough.

mhammond replied:
I pushed a fix to this.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1020 @@
+                self.store_bookmark(Staging::Incoming, timestamp, b)?
+            }
+            BookmarkItemRecord::Folder(f) => self.store_folder(Staging::Incoming, timestamp, f)?,
+            _ => unimplemented!("TODO: Store other types"),
```
Let's do this next...

mhammond replied:
This is now largely done, not withstanding a more detailed review of the dumb stuff I added :)

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1020 @@
+        self.merge_with_driver(&Driver)?;
+
+        let mut outgoing = OutgoingChangeset::new(self.collection_name().into(), inbound.timestamp);
+        // TODO: Fetch staged items from `itemsToUpload`.
```
And this. We'll end up closing out #728 in this PR, too.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1020 @@
+        new_timestamp: ServerTimestamp,
+        records_synced: &[String],
+    ) -> result::Result<(), failure::Error> {
+        unimplemented!("TODO: Write sync records back to the mirror and update sync statuses");
```
And also this. 😄

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1020 @@
+            json!({
+                "id": "qqVTRWhLBOu3",
+                "type": "bookmark",
+                "parentid": BookmarkRootGuid::Unfiled.as_guid(),
```
I _think_ this might be wrong. `parentid` (and `id`) should use [this format](https://searchfox.org/mozilla-central/rev/23fb3d3791eb9155fb31e76d322d6812cffded90/toolkit/components/places/PlacesSyncUtils.jsm#65-72) for roots, not the Places GUID format with underscores.

mhammond replied:
Nice catch - I updated the "incoming" tests to use the correct variant

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -391,9 +409,117 @@ impl<'a> BookmarksStore<'a> {
+        url: Url,
+    ) -> Result<(Option<Url>, SyncedBookmarkValidity)> {
+        // wow - this  is complex, but markh is struggling to see how to
+        // improve it
```
We could return `Result<(Url, SyncedBookmarkValidity)>` (and an error for invalid URLs), and have `store_incoming_query` assume than error means `Replace`, but that doesn't help flatten or simplify this.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -391,9 +409,117 @@ impl<'a> BookmarksStore<'a> {
+                    } else {
+                        let mut new_url = url.clone();
+                        new_url
+                            .query_pairs_mut()
```
Same here, we sadly can't use `query_pairs` directly.

mhammond commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -391,9 +409,117 @@ impl<'a> BookmarksStore<'a> {
+            // If the URL has `type={RESULTS_AS_TAG_CONTENTS}` then we
+            // rewrite the URL as `place:tag=...`
+            if url
+                .query_pairs()
```
I should have known that - thanks, fixed.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -515,6 +515,76 @@ impl<'a> BookmarksStore<'a> {
+
+    fn store_incoming_sep(&self, modified: ServerTimestamp, s: SeparatorRecord) -> Result<()> {
+        self.db.execute_named_cached(
+            r#"REPLACE INTO moz_bookmarks_synced(guid, parentGuid, serverModified, needsMerge, kind,
```
I was wondering if we should store the `position` from the record, too. On the one hand, newer Desktops (Dogear and the JS buffered engine) won't use it. On the other, it _might_ come in handy for fixing up orphaned separators—assuming their `parentid` exists—and we could still flag separators with the wrong `pos` for reupload.

Speaking of orphans, though...I wonder if it's better to just delete orphaned separators entirely, instead of trying to do fancy things with their `parentid` and `pos`. Moving separators to the end of the folder, or to `unfiled`, is unlikely to be useful, even if it is technically data loss.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            WHERE NOT v.isDeleted AND
+                  v.needsMerge AND
+                  b.guid IS NULL AND
+                  s.parentGuid <> '{root_guid}'"#,
```
`AND validity <> SyncedBookmarkValidity::Replace`, too; we can't meaningfully dedupe invalid items. (In fact, I think that will cause `fetch_new_remote_contents` to return an error now, since `url` will be null).

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            });
+        match result {
+            Ok(_) => tx.commit()?,
+            Err(_) => tx.rollback()?,
```
Not needed; `tx` will roll back automatically if it's not committed when dropped.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+
+        for incoming in inbound.changes {
+            applicator.apply_payload(incoming.0, timestamp)?;
+            incoming_telemetry.applied(1);
```
I left a note in 2382dc56414516fb7ca00e3d0226b1f7d7c58d5f, but I think we can record somewhat more informative stats here until we have event telemetry.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+                // We can't avoid allocating here, since we're binding four
+                // parameters per descendant. Rust's `SliceConcatExt::concat`
+                // is semantically equivalent, but requires a second allocation,
+                // which we _can_ avoid.
```
...which we _can_ avoid by writing this out.

@thomcc, can you think of any way to tighten this up?

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+        )?;
+
+        // Next, insert rows for deletions. Unlike Desktop, there's no
+        // `noteItemRemoved` trigger on `itemsToRemove`, since we don't support
```
This comment is more confusing than helpful here. Move to `create_sync_temp_tables.sql` or remove.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            Ok(())
+        })?;
+
+        // "Deleting" from `itemsToMerge` fires the `insertNewLocalItems` and
```
Explain air quotes (`itemsToMerge` is a view, so we don't really delete from it).

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            JOIN mergedTree r ON r.mergedGuid = b.guid
+            JOIN moz_bookmarks_synced v ON v.guid = r.remoteGuid
+            WHERE r.useRemote AND
+                  /* "b.dateAdded" is in microseconds; "v.dateAdded" is in
```
Comment out of date, they're both milliseconds.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            SELECT s.id, s.guid, s.syncChangeCounter, s.parentGuid,
+                   s.parentTitle, s.dateAdded, s.title, s.placeId,
+                   {kind}, h.url, NULL AS keyword, s.position,
+                   NULL AS tagFolderName
```
Drop this: we can't have an outgoing `tagFolderName`, because we rewrite tag queries on the way in, and can't make new queries locally.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            FROM localItems s
+            LEFT JOIN moz_places h ON h.id = s.placeId
+            LEFT JOIN idsToWeaklyUpload w ON w.id = s.id
+            WHERE s.syncChangeCounter >= 1 OR
```
Nit: `syncChangeCounter > 0`, to match `has_changes` and make grepping easier.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+        let result = self.push_synced_items(new_timestamp, records_synced);
+        match result {
+            Ok(_) => tx.commit()?,
+            Err(_) => tx.rollback()?,
```
No need to manually roll back.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+    }
+
+    fn get_collection_request(&self) -> result::Result<CollectionRequest, failure::Error> {
+        let since = get_meta::<i64>(self.db, LAST_SYNC_META_KEY)?
```
Desktop's logic for this is [a bit fancier](https://searchfox.org/mozilla-central/rev/f1c7ba91fad60bfea184006f3728dd6ac48c8e56/toolkit/components/places/SyncedBookmarksMirror.jsm#241-265), because it tries to handle the case where the sync was interrupted as we were downloading new items. It's only likely to come up for first syncs, or for clients that haven't synced in a while. For interrupted large batches, all records will have the same timestamp, so we'll still download them again.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+    }
+
+    fn reset(&self) -> result::Result<(), failure::Error> {
+        unimplemented!("TODO: Wipe staged items and reset sync statuses");
```
TODO!

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+    }
+}
+
+fn type_to_kind<'a>(typ: &'a str, url: UrlOrPlaceId<'a>) -> TypeToKind<'a> {
```
I learned this trick from @thomcc's `sql_support`. Maybe it needs a better name, though? `type_to_kind_fragment`?

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+
+        // And finally, a more "functional" test - that our queries end up
+        // correctly in the local tree.
+        // XXX - this test fails for reasons I don't understand. If we write
```
We talked about this on Slack, but, because we don't store `unfiled`, `query1______` is orphaned, so we'll delete it on the server. If you add a record for `unfiled` mentioning `query1______` in its children, it should apply.

mhammond replied:
yeah, thanks - `test_apply_query` has this test reinstated.

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+            .expect("Should fetch info for unfiled")
+            .unwrap();
+        // XXX - presumably this next line should use `info_for_unfiled`, but
+        // info_for_unfiled has a change counter of 1!
```
Oops, that's a bug! 🐞

linacambridge commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1233 @@
+        conn
+    }
+
+    fn assert_incoming_creates_local_tree(
```
Nice! :+1: 💯

linacambridge commented on components/places/src/tests.rs:
```diff
@@ -0,0 +1,258 @@
+// can correctly check Option<> fields (ie, so that None isn't ambiguous
+// between "no value specified" and "value is exactly None"
+#[derive(Debug)]
+pub enum MirrorBookmarkValue<T> {
```
Maybe `SyncedBookmarkValue`? 😉 I like this, though! 👍

mhammond commented on components/places/src/bookmark_sync/record.rs:
```diff
@@ -0,0 +1,348 @@
+}
+
+impl Serialize for BookmarkItemRecord {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
```
Yeah, I think this is fine

mhammond commented on components/places/src/bookmark_sync/record.rs:
```diff
@@ -0,0 +1,348 @@
+/// Converts a Sync bookmark record ID to a Places GUID. Sync record IDs are
+/// identical to Places GUIDs for all items except roots.
+#[inline]
+pub fn id_to_guid(id: String) -> SyncGuid {
```
I was chatting a little with Thom about this, and if we had a version of `id_to_guid_str(&str) -> &str` (and trivially built id_to_guid from that for convenience here) we could almost certainly avoid the string allocation in `push_synced_items`. However, that would involve duplicating some of the IDs though, I'm not sure it's worth it.

To your main point, these are the only callers of our extensions to `BookmarkRootGuid`, so I think we might as well consolidate them.

mhammond commented on components/places/sql/create_shared_temp_tables.sql:
```diff
@@ -44,3 +45,45 @@ CREATE TEMP TABLE moz_updateoriginsupdate_temp (
     PRIMARY KEY (prefix, host)
 ) WITHOUT ROWID;
 
+-- Tags
```
Why are these temp triggers now here? I'd expect to see them in `create_shared_triggers.sql`

linacambridge replied:
Yep, I pasted them into the wrong file! 📋 🍝

thomcc commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1723 @@
+
+/// A helper that interpolates a SQL expression for a Place URL. This avoids a
+/// subquery if the URL is already available in the projection.
+enum UrlOrPlaceId<'a> {
```
I think it would be easy for someone to assume this is supposed to represent either a Url or a Guid (or something), rather than what they are, which are column names. I can think of three ways to help avoid this:

1. Note in the comment that these are suppsed to be column references and not e.g. urls, or place IDs
2. Force these to be &'static str instead of parameterizing on lifetime. We should always be providing strings known at compile time (if tests need an escape hatch they can always go through a box and Box::leak the string to get a static str or something, but I'd be surprised if it was needed)
3. Maybe suffix types like this with `Expr` or `Fragment` to indicate they're effectively SQL expressions/fragments?

mhammond replied:
did *both* 'static and *Fragment!

thomcc commented on components/places/src/bookmark_sync/tests/synced_item.rs:
```diff
@@ -0,0 +1,218 @@
+use sync15::ServerTimestamp;
+use url::Url;
+
+// Our prod code never needs to read moz_bookmarks_synced, but our test code
```
Nit: `///`. This is clever though.

thomcc commented on components/places/src/bookmark_sync/tests/synced_item.rs:
```diff
@@ -0,0 +1,218 @@
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct SyncedBookmarkItem {
```
Oh boy that's a lot of fields.

thomcc commented on components/places/src/db/schema.rs:
```diff
@@ -174,6 +184,7 @@ fn upgrade(db: &PlacesDb, from: i64) -> Result<()> {
     )?;
     migration(db, 4, 5, &[CREATE_SHARED_SCHEMA_SQL], || Ok(()))?;
     migration(db, 5, 6, &[CREATE_SHARED_SCHEMA_SQL], || Ok(()))?; // new tags tables.
+    migration(db, 6, 7, &[CREATE_SHARED_SCHEMA_SQL], || Ok(()))?; // bookmark syncing.
```
I do feel as though 'rerun the whole schema' is going to eventually prove problematic, but we can cross that bridge then, I guess.

linacambridge replied:
Filed #849 so we don't lose track of this.

thomcc commented on components/support/sql/src/each_chunk.rs:
```diff
@@ -71,13 +71,13 @@ where
 ) -> Result<(), E>
 where
-    T: ToSql + 'a,
+    T: 'a,
```
Huh, right, that will make this more useful, won't it.

thomcc commented on components/places/src/types.rs:
```diff
@@ -44,6 +44,12 @@ where
 }
 
+impl From<SyncGuid> for dogear::Guid {
+    fn from(guid: SyncGuid) -> dogear::Guid {
```
I almost feel like we should just use dogear::Guid everywhere, although I don't actually that would work in practice.

thomcc commented on components/places/Cargo.toml:
```diff
@@ -10,7 +10,7 @@ default = []
 [dependencies]
 sync15 = { path = "../sync15" }
-serde = "1.0.75"
+serde = { version = "1.0.75", features = ["derive"] }
```
We shouldn't add derive and still depend on serde_derive.

Also lets not add this until we can add it globally, since it ends up slowing down compilation because everything depends on serde, and so turning on/off features will cause it to need to be recompiled, following with all things that depend on it.

thomcc commented on components/places/examples/places-utils.rs:
```diff
@@ -156,44 +158,92 @@ fn run_native_export(db: &PlacesDb, filename: String) -> Result<()> {
+    // XXX - unfortunately, history stores global meta in a `history_global_state`,
+    // but that's a global that should be shared between history and bookmarks.
+    // We should consider changing that key name?
+    // Even more unfortunate, places::storage::get_meta is `pub(crate)`, so we
```
We could change that and it wouldn't bother me at all. Probably better than having duplication here, tbh.

thomcc commented on components/places/sql/create_sync_triggers.sql:
```diff
@@ -0,0 +1,171 @@
+  -- Reduce the change counter and update the sync status for uploaded items.
+  -- If the item was uploaded during the sync, its change counter will still
+  -- be > 0 for the next sync.
+  UPDATE moz_bookmarks SET
```
Nit: We use 4 spaces to indent trigger bodies elsewhere so should here too IMO.

Also I don't seem to be able to see all my comments so sorry if I say this multiple times or whatever

thomcc commented on components/places/src/bookmark_sync/incoming.rs:
```diff
@@ -0,0 +1,646 @@
+                                                 dateAdded, title, keyword, validity, placeId)
+               VALUES(:guid, :parentGuid, :serverModified, 1, :kind,
+                      :dateAdded, NULLIF(:title, ""), :keyword, :validity,
+                      -- XXX - when url is null we still fail below when we call hash()???
```
Is this still true? I though we landed a fix for this...

thomcc commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1723 @@
+        // we don't store them yet, hence never "merged" (see bug 1343103).
+        let sql = format!(
+            "
+            SELECT
```
Nit: This query has weird indentation.

thomcc commented on components/places/sql/create_sync_triggers.sql:
```diff
@@ -0,0 +1,171 @@
+CREATE TEMP TRIGGER removeLocalItems
+AFTER DELETE ON itemsToRemove
+BEGIN
+  -- Flag URL frecency for recalculation.
```
Uh, do we actually recalculate negative frecencies anywhere? I also don't know that `frecency = -frecency` wont confuse the origin frecency update triggers badly...

linacambridge replied:
I added a `moz_places_stale_frecencies` table to hold Places with stale frecencies, but I haven't added logic to recalculate them yet. I wonder if we could, or should, do this on a background thread...or if we do this more like Desktop, where we backfill a fixed number (400) after every sync, or until maintenance runs, until we've recalculated everything.

thomcc commented on components/places/src/bookmark_sync/incoming.rs:
```diff
@@ -0,0 +1,646 @@
+                        }
+                    }
+                    Err(e) => {
+                        log::warn!(
```
god rustfmt makes log messages look so much worse.

mhammond commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1452 @@
+            // Note that this doesn't account for local clock skew.
+            let age = row
+                .get_checked::<_, Timestamp>("localModified")
+                .unwrap_or_default()
```
these fields are non-null, so no need to use Option<>, but I added the '?'

mhammond commented on components/places/src/bookmark_sync/store.rs:
```diff
@@ -0,0 +1,1452 @@
+            let guid = row.get_checked::<_, SyncGuid>("guid")?;
+            let kind = SyncedBookmarkKind::from_u8(row.get_checked("kind")?)?;
+            let mut item = Item::new(guid.into(), kind.into());
+            let age = ServerTimestamp(row.get_checked::<_, f64>("serverModified").unwrap_or(0f64))
```
this is also non nullable, but added the `?` to get_checked()

linacambridge commented on components/places/src/bookmark_sync/record.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+}
+
+impl Serialize for BookmarkItemRecord {
```
I 📋🍝 this from `storage::bookmarks`, but we totally can! Getting the record ID-GUID conversions right was tricky (I had to implement a visitor 🤨), but I think I've made it work, and it cuts down on the number of types.

linacambridge commented on components/places/src/bookmark_sync/record.rs:
```diff
@@ -0,0 +1,348 @@
+}
+
+#[derive(Debug, Clone, Hash, PartialEq)]
+pub struct BookmarkRecord {
```
At least we got rid of `RawBookmarkItemRecord`. :joy: (What a mouthful!)

linacambridge commented:
[Clippy complains](https://circleci.com/gh/mozilla/application-services/6535) the `serde_derive::*` import in `bookmark_sync/record.rs` is unused, but I definitely can't remove it, or it won't compile (`"The attribute `serde` is currently unknown to the compiler and may have meaning added to it in the future"`).

Barring that, frecency (https://github.com/mozilla/application-services/pull/723#discussion_r268016201, #847), and https://github.com/mozilla/application-services/pull/723#discussion_r268013171, I think this is ready to go.

thomcc commented:
> Clippy complains the serde_derive::* import in bookmark_sync/record.rs is unused, but I definitely can't remove it, or it won't compile ("The attribute serde is currently unknown to the compiler and may have meaning added to it in the future").

Hmm, we use serde in other files it's fine with. I think we just have to import Serialize and Deserialize explicitly

linacambridge commented:
I figured it out—all our other files import `serde::{de::Deserialize, ser::Serialize}`, but I was importing `serde::{Deserialize, Serialize}`. With that fixed, Clippy is happy. 📎 👍 

@thomcc, mind giving this a quick pass (if you haven't already!) before we land?
